### ISSUE TITLE
release a new version of kubeadm

### ISSUE THREAD

**Issue by luxas** [kubernetes/kubernetes#34884]
The last two weeks I've been focusing on getting kubeadm master more stable so we can do a release, and implementing new features. I'm collecting here now what should be done:

Changelog for the new release:
- Switch to the 10.12.0.0/16 subnet: https://github.com/kubernetes/kubernetes/pull/33668
- Fix kubeadm on AWS by including /etc/ssl/certs in the controller-manager https://github.com/kubernetes/kubernetes/pull/33681
- The API was refactored and is now componentconfig: https://github.com/kubernetes/kubernetes/pull/33728, https://github.com/kubernetes/kubernetes/pull/34147 and https://github.com/kubernetes/kubernetes/pull/34555
- Allow kubeadm to get config options from a file: https://github.com/kubernetes/kubernetes/pull/34501, https://github.com/kubernetes/kubernetes/pull/34885 and https://github.com/kubernetes/kubernetes/pull/34891
- Implement preflight checks: https://github.com/kubernetes/kubernetes/pull/34341
- Using kubernetes v1.4.1 by default for arm support: https://github.com/kubernetes/kubernetes/pull/34419
- Make api and discovery ports configurable and default to 6443: https://github.com/kubernetes/kubernetes/pull/34719
- Implement kubeadm reset: https://github.com/kubernetes/kubernetes/pull/34807
- Make kubeadm poll/wait for endpoints instead of directly fail when the master isn't available https://github.com/kubernetes/kubernetes/pull/34703 and https://github.com/kubernetes/kubernetes/pull/34718
- Bug fixes: https://github.com/kubernetes/kubernetes/pull/34352, https://github.com/kubernetes/kubernetes/pull/34558, https://github.com/kubernetes/kubernetes/pull/34573, https://github.com/kubernetes/kubernetes/pull/34834 and https://github.com/kubernetes/kubernetes/pull/34607

My plan is to get https://github.com/kubernetes/kubernetes/pull/34718, https://github.com/kubernetes/kubernetes/pull/34719 and https://github.com/kubernetes/kubernetes/pull/34807 merged under the weekend so we can do the release on Monday.

Now that https://github.com/kubernetes/test-infra/issues/670 is fixed, we will also be able to just download the kubeadm binaries for all arches right away when building the debs and rpms.
I'll make a PR when we have a commit version that includes all above PRs.

I'd like to document this changelog somewhere, where do you think it would fit in the best?

@kubernetes/sig-cluster-lifecycle

**Comment by errordeveloper** [comment]
Yes, I also would like to fix https://github.com/kubernetes/kubernetes/issues/34927.

**Comment by luxas** [comment]
We should aim to get the release out of the door today. 
I think we'll manage without #34927, but #34719 and #34907 need to get merged.

Please fix the rebase failure and I'll merge it.

**Comment by pires** [comment]
> I think we'll manage without #34927, but #34719 and #34907 need to get merged.

Agreed.

**Comment by errordeveloper** [comment]
Ok, turns out I've used an old build when I created #34927. Pre-flight check errors look ok when I build from master.

**Comment by errordeveloper** [comment]
Now that we have all PR merged, we need to agree on the process. First of all we need to release latest binaries into the unstable channel, so we can test end-to-end from packages. I believe we have a URL for binaries built in CI, who know what it is should make a PR to update [the URL](https://github.com/kubernetes/release/blob/05117a866c69619f426045bc05b01da12b7548b0/debian/xenial/kubeadm/debian/rules#L13) first.

**Comment by pires** [comment]
> we need to release latest binaries into the unstable channel, so we can test end-to-end from packages.

I definitely agree we must test all scenarios manually before releasing. If we can’t, we shouldn’t release.

**Comment by errordeveloper** [comment]
cc @mikedanese @dgoodwin

**Comment by luxas** [comment]
The packaging scripts in kubernetes/release now uses the latest version: https://github.com/kubernetes/release/pull/160

When you've verified they are working properly, @mikedanese should push them to unstable and stable

**Comment by pires** [comment]
@dmmcquay please test as well and provide your feedback here.

**Comment by mikedanese** [comment]
@luxas debs are pushed automatically to stable. We need to pick a point to tag unstable as new stable.

**Comment by mikedanese** [comment]
I think we should pick the v1.5-alpha.2 release for this, not a ci build off master. We should be tapping into the main kubernetes release cycle now that we are in the main repository.

**Comment by mikedanese** [comment]
@saad-ali when are you planning on cutting v1.5-alpha.2 ?

**Comment by luxas** [comment]
v1.5.0-alpha.1 just got released, so we should **not** wait ~2 weeks for the next release for this.

This time a commit is ok, but for the next time, sure

**Comment by mikedanese** [comment]
According to the release calendar there should be a v1.5-alpha.2 today. I think v1.5-alpha.1 slipped a week. Is that correct @saad-ali ?

**Comment by mikedanese** [comment]
Oops misread. Looks like next one is on 10/31.

**Comment by saad-ali** [comment]
> @saad-ali when are you planning on cutting v1.5-alpha.2 ?

`v1.5.0-alpha.1` went out on Oct 13. Keeping a two week cadence, `v1.5.0-alpha.2` is planned for the 27th

**Comment by errordeveloper** [comment]
I believe test instructions for someone who has already installed kubeadm would be something like:


### On Ubuntu 16.04
```
# cat <<EOF > /etc/apt/sources.list.d/kubernetes.list
deb http://apt.kubernetes.io/ kubernetes-xenial-unstable main
EOF
# apt-get update
# apt-get upgrade -y kubelet kubeadm kubectl kubernetes-cni
```

## On CentOS or other EL7-based distros
```
# cat <<EOF > /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=http://yum.kubernetes.io/repos/kubernetes-el7-x86_64-unstable
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg
       https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
EOF
# yum upgrade -y docker kubelet kubeadm kubectl kubernetes-cni
```

And finally:
```
# kubeadm reset
# systemctl restart kubelet
# kubeadm init
```

Can anyone here test this, so we can can point folks who had issue at these?

**Comment by thaume** [comment]
Hey @errordeveloper thanks for the update ! Just ran your script, here are the results.

After updating the master, the other node disapeared (I guess this is pretty normal, since the `kubeadm reset` and `kubeadm init` create a new "join" token).
```
NAME       STATUS    AGE
sd-83688   Ready     40s
```

After running the update on the second pod + joining, my apiserver and k8s dashboard are fiercly running on port `:6443`.

A quick note, a preflight test outputed that : 
```
preflight check errors:
	ethtool not found in system path
```
It was an `apt-get install` away, but it validates that preflights are working :P

Thanks again for the quick update !

**Comment by Starefossen** [comment]
@errordeveloper are you sure the new packages are pushed to `-unstable`?

The `filelists.xml` looks very much the same:

* https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64/repodata/filelists.xml
* https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64-unstable/repodata/filelists.xml

**Comment by pesho** [comment]
Only the DEB packages have been pushed: https://packages.cloud.google.com/apt/dists/kubernetes-xenial-unstable/main/binary-amd64/Packages

**Comment by errordeveloper** [comment]
It is possible there is no automation for RPMs yet.

**Comment by mikedanese** [comment]
@Starefossen try again in unstable now.

@errordeveloper @dgoodwin @luxas I'm ready to tag unstable as stable. Just waiting for the final word.

**Comment by dgoodwin** [comment]
I did some manual testing with latest unstable rpms, bring up cluster and join nodes, everything working as expected on CentOS 7.

**Comment by errordeveloper** [comment]
I think @pires has done fair amount of testing also, I'll re-test on Ubuntu and ping back once done.

**Comment by Starefossen** [comment]
@mikedanese thanks mate!

Just fyi, do a `sudo yum clean metadata` before upgrade to make the `unstable` channel kick in.

**Comment by Starefossen** [comment]
I am getting the following error from my `kube-dns` with new `kubeadm`:

```
$ kubectl log kube-dns-2247936740-atkxb -c kube-dns -n kube-system

E1020 18:48:57.611180       1 reflector.go:214] pkg/dns/dns.go:155: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: getsockopt: no route to host
F1020 18:48:57.704372       1 server.go:133] Received signal: terminated
```

**Comment by pesho** [comment]
I'm also having DNS issues on a test cluster set up from scratch using the "unstable" `kubeadm`. The nameserver IP which pods receive in their `/etc/resolv.conf` (10.12.0.10) does not match the IP of the `kube-dns` service (10.0.0.10). As a result all DNS lookups within pods fail.

**Comment by webwurst** [comment]
Same problem like @pesho. See over there for details:
https://gist.github.com/webwurst/e65839c4889b8c3c88051ffd7b072168

**Comment by n-marton** [comment]
The generation of (on ubuntu, didn't check on other dists.) /etc/systemd/system/kubelet.service.d/10-kubeadm.conf ignores  custom  --service-cidr and contains 10.12.0.10 as --cluster-dns.

**Comment by errordeveloper** [comment]
I am suspecting there may be an issue with recent change that went in with fbd5032da26a3c014e1be48490a0c2ca6ee21c77.

**Comment by errordeveloper** [comment]
Having said that, I am not happy with `10.12.0.0/12` being the default range, it's too likely to clash, for example there will be a major overlap with `10.0.0.0/16`. I'd like to change the default.

**Comment by errordeveloper** [comment]
https://github.com/kubernetes/release/pull/169

**Comment by Starefossen** [comment]
Applied the dns fix on my kubeadm cluster but the errors still presist from within the pods:

Curling the kube-apiserver from a node:

```
[root@node-01 ~]# curl -k https://10.0.0.1
Unauthorized
```

Curling the kube-apiserver from within a pod:

```
[root@node-01 ~]# kubectl exec test-701078429-ael0b -- curl --verbose -k https://10.0.0.1
curl: (7) Failed to connect to 10.0.0.1 port 443: No route to host
```

```
[root@node-01 ~]# kubectl exec test-701078429-ael0b -- cat /etc/resolv.conf 
nameserver 10.0.0.10
```

```
[root@node-01 ~]# kubectl get svc --all-namespaces
NAMESPACE     NAME         CLUSTER-IP   EXTERNAL-IP   PORT(S)         AGE
default       kubernetes   10.0.0.1     <none>        443/TCP         23m
kube-system   kube-dns     10.0.0.10    <none>        53/UDP,53/TCP   22m
```

**Comment by errordeveloper** [comment]
@Starefossen this looks like a completely separate issue, let's discuss on Slack.

**Comment by errordeveloper** [comment]
I am waiting for at least https://github.com/kubernetes/release/pull/168 & https://github.com/kubernetes/kubernetes/pull/35270 to get meged, also https://github.com/kubernetes/kubernetes/pull/35290 would be good.

**Comment by vganapathy1** [comment]
@errordeveloper  I'm also facing the same issue as @Starefossen, can you please suggest?

**Comment by Starefossen** [comment]
@vagababov I managed to get my kubernetes cluster working with internal and external dns queries on Oracle Linux 7.2 (kernel `4.1.12-61.1.14.el7uek.x86_64`) after a complete reset:

```

**Comment by errordeveloper** [comment]
@vganapathy1 is it possible you might have some firewall configuration that gets on the way? What distro are you using?

**Comment by errordeveloper** [comment]
Looks like we are also blocked on https://github.com/kubernetes/release/issues/171.

**Comment by vganapathy1** [comment]
@errordeveloper, I using Ubuntu 16.04.1 LTS (GNU/Linux 4.4.0-43-generic x86_64) and firewall is by default disabled. I even tried to reset steps suggested by @Starefossen and still the issue persists!

FYI When i look at the kube-proxy-adm64 I could see the below messages,
I1025 11:48:32.834269       1 iptables.go:339] running iptables-restore [--noflush --counters]
I1025 11:48:32.837936       1 proxier.go:751] syncProxyRules took 21.882689ms
I1025 11:48:32.837961       1 proxier.go:523] OnEndpointsUpdate took 22.098958ms for 4 endpoints
I1025 11:48:33.107442       1 config.go:99] Calling handler.OnEndpointsUpdate()
I1025 11:48:33.107485       1 proxier.go:758] Syncing iptables rules
I1025 11:48:33.107496       1 iptables.go:362] running iptables -N [KUBE-SERVICES -t filter]
I1025 11:48:33.108745       1 healthcheck.go:86] LB service health check mutation request Service: default/kubernetes - 0 Endpoints []
I1025 11:48:33.109389       1 iptables.go:362] running iptables -N [KUBE-SERVICES -t nat]
I1025 11:48:33.111039       1 iptables.go:362] running iptables -C [OUTPUT -t filter -m comment --comment kubernetes service portals -j KUBE-SERVICES]
I1025 11:48:33.113422       1 iptables.go:362] running iptables -C [OUTPUT -t nat -m comment --comment kubernetes service portals -j KUBE-SERVICES]
I1025 11:48:33.115332       1 iptables.go:362] running iptables -C [PREROUTING -t nat -m comment --comment kubernetes service portals -j KUBE-SERVICES]
I1025 11:48:33.117222       1 iptables.go:362] running iptables -N [KUBE-POSTROUTING -t nat]
I1025 11:48:33.118631       1 iptables.go:362] running iptables -C [POSTROUTING -t nat -m comment --comment kubernetes postrouting rules -j KUBE-POSTROUTING]
I1025 11:48:33.120569       1 iptables.go:298] running iptables-save [-t filter]
I1025 11:48:33.123196       1 iptables.go:298] running iptables-save [-t nat]
I1025 11:48:33.125567       1 proxier.go:1244] Restoring iptables rules: *filter
:KUBE-SERVICES - [0:0]
-A KUBE-SERVICES -m comment --comment "kube-system/kube-dns:dns has no endpoints" -m udp -p udp -d 10.0.0.10/32 --dport 53 -j REJECT
-A KUBE-SERVICES -m comment --comment "kube-system/kube-dns:dns-tcp has no endpoints" -m tcp -p tcp -d 10.0.0.10/32 --dport 53 -j REJECT
COMMIT
*nat
:KUBE-SERVICES - [0:0]
:KUBE-NODEPORTS - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-MARK-MASQ - [0:0]
:KUBE-SVC-NPX46M4PTMTKRN6Y - [0:0]
:KUBE-SEP-TKCQEPMSAIXROJ4U - [0:0]
:KUBE-SVC-TCOU7JCQXEZGVUNU - [0:0]
:KUBE-SVC-ERIFXISQEP7F7OF4 - [0:0]
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -m mark --mark 0x00004000/0x00004000 -j MASQUERADE
-A KUBE-MARK-MASQ -j MARK --set-xmark 0x00004000/0x00004000
-A KUBE-SERVICES -m comment --comment "default/kubernetes:https cluster IP" -m tcp -p tcp -d 10.0.0.1/32 --dport 443 -j KUBE-SVC-NPX46M4PTMTKRN6Y
-A KUBE-SVC-NPX46M4PTMTKRN6Y -m comment --comment default/kubernetes:https -m recent --name KUBE-SEP-TKCQEPMSAIXROJ4U --rcheck --seconds 180 --reap -j KUBE-SEP-TKCQEPMSAIXROJ4U
-A KUBE-SVC-NPX46M4PTMTKRN6Y -m comment --comment default/kubernetes:https -j KUBE-SEP-TKCQEPMSAIXROJ4U
-A KUBE-SEP-TKCQEPMSAIXROJ4U -m comment --comment default/kubernetes:https -s 10.63.33.46/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-TKCQEPMSAIXROJ4U -m comment --comment default/kubernetes:https -m recent --name KUBE-SEP-TKCQEPMSAIXROJ4U --set -m tcp -p tcp -j DNAT --to-destination 10.63.33.46:6443
-A KUBE-SERVICES -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp -p udp -d 10.0.0.10/32 --dport 53 -j KUBE-SVC-TCOU7JCQXEZGVUNU
-A KUBE-SERVICES -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp -p tcp -d 10.0.0.10/32 --dport 53 -j KUBE-SVC-ERIFXISQEP7F7OF4
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
COMMIT

**Comment by errordeveloper** [comment]
Does 'curl -vk https://10.0.0.1:443' work from a node and master?

**Comment by luxas** [comment]
Please move the debug session to another thread so we can focus on releasing tasks here instead :)

**Comment by errordeveloper** [comment]
Ok, looks like there is a bug, I wonder why we missed it so far, it may be
an edge case. Tobias, if you can give me access, it would be the easiest
path to debug this. I wonder why the /etc/kubernetes directory was not
empty in your VM... Before I get to this, could you check what was in it,
and in case there may be any other state, e.g. data in etcd?

**Comment by errordeveloper** [comment]
Let's release another unstable snapshot ASAP https://github.com/kubernetes/release/pull/177 👍

**Comment by errordeveloper** [comment]
> Please move the debug session to another thread so we can focus on releasing tasks here instead :)

Thanks @luxas. @vganapathy1 if you are still having problems, please open another issue or find me on Slack.

**Comment by luxas** [comment]
Some more PRs will be in the next release:
 - The preflight checks do now ensure that the hostname of the node is allowed #35843
 - Default service CIDR is now 10.96.0.0/12 #35290
 - v1.4.4 of kubernetes is used by default #35270
 - Added support for config file for kubeadm join as well #34885
 - Allow empty directories in the directory preflight check: #35632
 - Bug fixes: #34907 #34891 #35796
 - Various enhancements: #35075 #35111 #35119 #35124 #35265 #35777
 - Started adding unit tests: #35231 #35326 #35332 

The last piece we need is #35796 before we can push the release to stable in my opinion.

cc @pires @errordeveloper @mikedanese

**Comment by bulletRush** [comment]
in kubeadm source code, some place use "k8s.io/kubernetes/pkg/api" and some place use "k8s.io/kubernetes/pkg/api/v1".  and `v1.PodSecurityContext` doesn't have `HostNetwork` field. 
find more information in my branch: [pre pull images and configurable pod](https://github.com/bulletRush/kubernetes/commit/7d73b13bd4a4b346dd13fd0d6808a3a49e37110a)

**Comment by luxas** [comment]
Now all required PRs are merged in order to produce a new stable release.
https://github.com/kubernetes/release/pull/186 promotes the current HEAD version to unstable, and after that we're gonna mark it as stable as well.

Soon a docs update is coming

**Comment by luxas** [comment]
Docs PR: https://github.com/kubernetes/kubernetes.github.io/pull/1608#issue-186868008

**Comment by dgoodwin** [comment]
I have found a bug with kubeadm join on CentOS, we fixed master init, we failed to think about join:

```
$ kubeadm join --token=736a96.7d67243bc4d6137a 192.168.122.176
Running pre-flight checks
preflight check errors:
        kubelet service is not active, please run 'systemctl start kubelet.service'
        /etc/kubernetes is not empty
```

Fixing now.

**Comment by luxas** [comment]
Oops, yeah. We need tests :sweat_smile: 
I'm merging that one now

**Comment by luxas** [comment]
@mikedanese Please promote the unstable packages to stable today.
I expect it to take about ~5 mins for you, so I think I'm under the 2 hour limit :smile:

**Comment by luxas** [comment]
The release is out now and documented, closing...

**Comment by sanketmlad** [comment]
Is there any plan to upgrade kubeadm so that Kubernetes cluster will handle load balancing by themselves and will kube-apiserver will not require external load balancing?
